### PR TITLE
Add config file watch service

### PR DESCRIPTION
**Pull request by K1li4nL**
This pr adds the ability for the system to react on the fly to changes in the config files added in #1519. 

Limitations: Without greet_server implementation we are still susceptible to having double connections (A to B and B to A, over two websockets).

### CODE REVIEW

File: be2-scala/src/main/scala/ch/epfl/pop/decentralized/Monitor.scala [hunk 1]
```diff
@@ -86,3 +94,35 @@ object Monitor {
...
+      while (!Thread.currentThread().isInterrupted) {
+        // Blocks until an event happen
+        val watchKey = watchService.take()
+
+        // For any event, read the file and send it
+        for (event <- watchKey.pollEvents().asScala.toList) {
+          if (serverPeersListPath.endsWith(event.context().toString)) {
+            mediatorRef ! ConnectionMediator.ConnectTo(readServerPeers())
+          }
+          watchKey.reset()
```
Comment (Jharaxus) [inline be2-scala/src/main/scala/ch/epfl/pop/decentralized/Monitor.scala:118]: Do you need to do that? The events are already fetched and the watchKey will be deleted by the end of this while loop's scope anyway...

File: be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala [hunk 2]
```diff
@@ -99,4 +104,40 @@ class MonitorSuite extends TestKit(ActorSystem("MonitorSuiteActorSystem")) with
...
+    mockConnectionMediator.send(monitorRef, ConnectionMediator.Ping())
+
+    // Expect no message as long as the server peers list is untouched
+    mockConnectionMediator.expectNoMessage(timeout)
+
+    val newContent = List("some", "strings")
+    testWriteToServerPeersConfig(newContent)
+
+    mockConnectionMediator.expectMsgType[ConnectionMediator.ConnectTo](timeout)
+    deleteTestConfig()
```
Comment (Jharaxus) [inline be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala:122]: I don't think the test config file will be deleted if the test failed in the previous assertion

**Review by Jharaxus — CHANGES_REQUESTED** [review]
Good job, the tests of `FileMonitor` are very nice !

File: be2-scala/src/main/scala/ch/epfl/pop/decentralized/Monitor.scala (hunk 1 above)
Comment (K1li4nL) [reply be2-scala/src/main/scala/ch/epfl/pop/decentralized/Monitor.scala:118]: As I understood it, this line is mandatory to put the key back from `signalled` into the `ready` state. Even if the variable holding the ref goes out of scope, the key it points to seems to be living as long as the watch service is not cancelled. I am very not comfortable with this... Here is the doc:

File: be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala (hunk 2 above)
Comment (K1li4nL) [reply be2-scala/src/test/scala/ch/epfl/pop/decentralized/MonitorSuite.scala]: Fixed !

File: be2-scala/src/main/scala/ch/epfl/pop/decentralized/Monitor.scala (hunk 1 above)
Comment (Jharaxus) [reply be2-scala/src/main/scala/ch/epfl/pop/decentralized/Monitor.scala:118]: Ok that makes more sense now!
However I think this should be moved out of the `for` loop then

Comment (K1li4nL) [reply be2-scala/src/main/scala/ch/epfl/pop/decentralized/Monitor.scala:118]: I hadn't seen that I think... yes indeed

**Review by Jharaxus — APPROVED** [review]

**Review by Ajkunas — APPROVED** [review]
LGTM

**Pull request merged**